Associate saved card with the customer

luizreis opened the pull request:
Fixes #965

#### Changes proposed in this Pull Request

* Add customer ID to token's meta when creating it
* Do not display tokens that do not belong to current customer
* Add customer information to old tokens based on account mode and created customers
* Do not delete tokens that belong to the inactive customer

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. **Test accounts**
    1.1. Using a test account, add a card while using `trunk`
    1.2. Apply this PR and add a new card
    1.3. Assert that both tokens are displayed during checkout
    1.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
2. **Live accounts in test mode without live customer**
    3.1. Using a live account in test mode, add a card while using `trunk`
    3.2. Apply this PR and add a new card
    3.3. Assert that both tokens are displayed during checkout
    3.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
3. **Live accounts in test mode with live customer**
    3.1. Ensure your store has a live-mode customer before switching test mode on
    3.2. Using the live account in test mode, add a card while using `trunk`
    3.3. Apply this PR and add a new card
    3.4. Assert that only the new token is displayed during checkout
    3.5. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated to the customer from live mode
    3.6. Switch back to live mode, you should see the token from `3.2` in the checkout (even though it won't work)
4. **Live accounts in live mode**
    4.1. Using a live account in live mode, add a card while using `trunk`
    4.2. Apply this PR and add a new card
    4.3. Assert that both tokens are displayed during checkout
    4.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
    4.5. Switch the store to test mode and add a new card
    4.6. Assert that a new customer was created under the new test key
    4.7. Assert that only the test token is displayed during checkout
    4.8. Assert that the new token has the test customer in the `_wcpay_customer_id` meta

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

luizreis committed:
Add customer ID to token when creating it

luizreis committed:
Extract payment method import logic to import_customer_tokens

luizreis committed:
Remove unavailable tokens before displaying them

Tokens attached to a user might belong to different customers, for
example, live and test mode ones. This commit filters the gateway
token list and displays only the tokens that belong to the current
customer, to avoid issues during checkout.

luizreis committed:
Migrate tokens that have no customer to live mode

This commit adds the current customer ID to tokens that don't have
one in woocommerce_get_customer_payment_tokens, when in live mode.

luizreis committed:
Add changelog entry

kalessil commented:
Hey @DanReyLop: I recall 2 recent PRs from you were also dealing with payment methods syncronization. Is this PR scope intersecting with your changes somehow?

luizreis commented:
> Hey @DanReyLop: I recall 2 recent PRs from you were also dealing with payment methods synchronization. Is this PR scope intersecting with your changes somehow?
> 
> Edit: asking as trying to understand the context better

I think that's a great question! Especially when considering network-wide saved cards and customers.

DanReyLop commented:
I don't think this clashes with my PRs. The original behaviour of the tokens in WCPay was to fetch the payment methods from Stripe for a given customer, and if there are new ones, add them to the `Tokens` WC storage. The only change I made was to make sure to "purge" the tokens that don't exist in Stripe's side anymore, so Stripe is the main source of truth.

However, your change refactors the `woocommerce_get_customer_payment_tokens` function, so whoever merges his PR last will have to deal with some (easy, I hope) conflicts.

dwainm commented:
This doesn't seem to work for me. Here is `master`:

<img width="497" alt="Screenshot 2020-11-05 at 15 10 32" src="https://user-images.githubusercontent.com/1713474/98245198-3ecc1c80-1f79-11eb-8441-2a13a8e4c0ae.png">


Here's `origin/add/965-associate-saved-card-with-customer` after a branch switch and page reload.

<img width="514" alt="Screenshot 2020-11-05 at 15 10 41" src="https://user-images.githubusercontent.com/1713474/98245265-560b0a00-1f79-11eb-8fad-d602d715b5f1.png">

dwainm commented:
My accounts page shows the cards in master and this branch:

<img width="980" alt="Screenshot 2020-11-05 at 15 14 56" src="https://user-images.githubusercontent.com/1713474/98245527-b8fca100-1f79-11eb-9b09-62a969b83698.png">

luizreis commented:
Hey @dwainm, I think it would be better to review #971 first, as this PR builds on it. Were you using `test` mode to test this? If so, this is the expected behavior since we're migrating the customer and their tokens to `live` mode, as discussed in paJDYF-QQ-p2. I'll add that the testing instructions.

dwainm commented:
@luizreis Yes, I thought this PR includes code for both so I started here. Also, the live mode requirement on #972 is causing me to fix up our dev platform account to allow me to create live accounts, so that's my next step before testing it.

On this one, I used `test` mode. I see the migration logic is only for live mode, but wasn't sure why. Why can't we also migrate test mode card data to the new test mode key? I think having this work smoothly in `test` mode was my first expectation.

luizreis commented:
> @luizreis Yes, I thought this PR includes code for both so I started here. Also, the live mode requirement on #972 is causing me to fix up our dev platform account to allow me to create live accounts, so that's my next step before testing it.
> 
> On this one, I used `test` mode. I see the migration logic is only for live mode, but wasn't sure why. Why can't we also migrate test mode card data to the new test mode key? I think having this work smoothly in `test` mode was my first expectation.

There's more context in paJDYF-QQ-p2, but we cannot assume that the store mode is the mode in which the object was created. So, to avoid having a complicated migration strategy that we'd have to maintain, we decided to migrate everything to `live` mode to avoid losing real customer data if the migration happens when the store is in `test` mode.

Taking a look at paJDYF-QQ-p2#comment-3618 it seems like you agreed with it at the time. Do you think we should strive to support `test` mode migration now, even if it means we could end up with a PR as big and complicated as #918?

dwainm commented:
> “if the data exists under the new live key, return it, otherwise look for the legacy key. If a new test key was requested and it doesn’t exist, always return nothing”.

I think my understanding is mixed up with how we're supposed to handle the fallback. I thought we'd fall back to the legacy in test mode case too, but thinking about it more I realize that the legacy could contain live data, which is a bigger probability than it containing test keys. Sorry for causing confusion. 

As far as the test mode goes this work. I'll try and set up a live account tomorrow and test more around switching on both PR's.

luizreis commented:
> > “if the data exists under the new live key, return it, otherwise look for the legacy key. If a new test key was requested and it doesn’t exist, always return nothing”.
> 
> I think my understanding is mixed up with how we're supposed to handle the fallback. I thought we'd fall back to the legacy in test mode case too, but thinking about it more I realize that the legacy could contain live data, which is a bigger probability than it containing test keys. Sorry for causing confusion.

Gotcha, no worries 😄

> As far as the test mode goes this works well. I'll try and set up a live account tomorrow and test more around switching on both PR's.

Fantastic! Thank you!

dwainm commented on includes/class-wc-payments-token-service.php:
```diff
@@ -17,6 +17,8 @@
  */
 class WC_Payments_Token_Service {
 
+	const CUSTOMER_ID_META_KEY = '_wcpay_customer_id';
```
It'll be good to also call this `DEPRECATED_WCPAY_CUSTOMER_ID_OPTION` as it is in the other PR.

luizreis replied:
This specifies the meta key in which the customer ID is going to be attached to the token. It's the new default behavior to attach the customer to the token to avoid displaying tokens that will not be able to be used during checkout. Did you mean that we should name it `WCPAY_CUSTOMER_ID_OPTION`?

If so, I'd rather keep it as `META_KEY` due to the way it's used. The `CUSTOMER_ID_OPTION` ones are used to deal with user options using the `(get|update|delete)_user_option` functions, while this one is used to handle the token's metadata using the functions `(add|update)_meta_data`, `get_meta ` and `meta_exists`. Does that make sense?

dwainm replied:
Cleared out in Slack, the name made sense pasting Luiz's explanation here for future comment readers:


`_wcpay_customer_id` in the Token_Service relates to the wcpay_customer_id for that token, while `_wcpay_customer_id_live` and `_wcpay_customer_id_test` relate to the `wcpay_customer_id` attached to the WooCommerce user

The `wcpay_customer_id` is our Stripe customer entity. It has a relationship with the WC_User in the sense that we create Stripe customers for the WC_Users, but it also has a relationship with the token, as a Stripe user owns a Payment Method, which is stored in the token to enable saved cards and subscriptions.

luizreis replied:
I've added some comments to clarify that as well in https://github.com/Automattic/woocommerce-payments/pull/972/commits/d26d052f13925083fceb179bb012611748822b9e and https://github.com/Automattic/woocommerce-payments/pull/973/commits/aa26d7959aabd9b77583bf9a2765723dc2157d03.

luizreis committed:
Add clarification comments to the Token_Service constant

The constant name and meta key that we store a customer_id are not
very clear and have some clashes with constants that might be used
elsewhere. This commit adds some coments to clarify its usage.

luizreis commented:
Just noting here that we'll have to work around https://github.com/Automattic/woocommerce-payments/pull/1004 as it will delete any WC Pay token that's not assigned to the current customer. Once we start handling two customers per user, we should not remove tokens that belong to any of them, even if they'd be recreated once the store changes mode, to avoid issues when renewing subscriptions.

luizreis committed:
Add WC_Payments_Account to WC_Payments_Token_Service dependencies

luizreis committed:
Add optional $mode parameter to get_customer_id_by_user_id

luizreis committed:
Use account is_live instead of test_mode when migrating tokens

We can safely assume that an account that's not live will only have
test customers. So, we don't have to take the live mode optimistic
approach everytime. This avoids the loss of test customer data for
test accounts. The optimistic live mode approach is still used when
the account is_live.

luizreis committed:
Do not remove tokens that belong to the inactive customer

Stores can have two simultaneous customers, one for live and one
for test mode. This commit changes the code that removes methods
that no longer exist in Stripe to not remove methods that were
assigned to the inactive mode customer.

luizreis committed:
Fix token migration tests

luizreis committed:
Migrate live account tokens to test customer

Migrate live account tokens to test customer if a live one does not
exist. If a live customer does not exist, we can safely migrate the
tokens to the test mode customer.

dwainm commented:
@dechov @vbelolapotkov this could use a review for Sprint 23.

vbelolapotkov commented on includes/class-wc-payments-token-service.php:
```diff
@@ -133,46 +208,73 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+		}
+
+		foreach ( $tokens as $token ) {
+			if ( ! $token->meta_exists( self::CUSTOMER_ID_META_KEY ) ) {
```
As I understand `$tokens` might have other gateway tokens as well (e.g. Stripe). Would it make sense to check if `$tokent->get_gateway_id() === WC_Payment_Gateway_WCPay::GATEWAY_ID`? We do this [here](https://github.com/Automattic/woocommerce-payments/blob/c2d42e9bb505eb8498c5647f85a9f4090041e2c7/includes/class-wc-payments-token-service.php#L188-L188).

luizreis replied:
Good catch, it does! Fixed in https://github.com/Automattic/woocommerce-payments/pull/973/commits/a625b21f42fb1799af0eef27626f5dbb7e4eab91.

vbelolapotkov commented on tests/test-class-wc-payments-token-service.php:
```diff
@@ -197,17 +215,39 @@ public function test_woocommerce_payment_token_set_default_no_customer() {
+		$token2->add_meta_data( '_wcpay_customer_id', 'cus_12345' );
+		$unavailable_token->add_meta_data( '_wcpay_customer_id', 'cus_67890' );
+
+		$tokens = [
```
It looks like `$tokens` passed to the WC filter might come from another gateways too. Let's add other gateway token to make sure it's not filtered out by our pipeline.

luizreis replied:
Good catch, added in https://github.com/Automattic/woocommerce-payments/pull/973/commits/0ff2e7135db57e1cc80c4e6f02d5d26dd216efc8 and https://github.com/Automattic/woocommerce-payments/pull/973/commits/a625b21f42fb1799af0eef27626f5dbb7e4eab91.

vbelolapotkov reviewed (commented):
@luizreis, sorry but I didn't have time to properly test it with live account. Left some thoughts on the code only. It works well with migrating test mode tokens. Besides that everything looks good. So I feel it can be shipped after more thorough testing.

luizreis committed:
Improve clarity for the $inactive_customer_id variable

luizreis committed:
Check gateway before removing unavailable token

luizreis committed:
Check gateway before migrating old tokens

luizreis commented on includes/class-wc-payments-token-service.php:
```diff
@@ -133,46 +208,73 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+		return array_filter(
+			$tokens,
+			function ( $token ) use ( $customer_id ) {
+				return $token->get_meta( self::CUSTOMER_ID_META_KEY ) === $customer_id;
```
Yep, fixed in https://github.com/Automattic/woocommerce-payments/pull/973/commits/0ff2e7135db57e1cc80c4e6f02d5d26dd216efc8.

luizreis commented on includes/class-wc-payments-token-service.php:
```diff
@@ -133,46 +208,73 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
 
 		// Remove the payment methods that no longer exist in Stripe's side.
 		remove_action( 'woocommerce_payment_token_deleted', [ $this, 'woocommerce_payment_token_deleted' ], 10, 2 );
+		$inactive_customer_id = $this->customer_service->get_customer_id_by_user_id(
```
Sure! What do you think of https://github.com/Automattic/woocommerce-payments/pull/973/commits/462a8face7d542bc6f90807e358782fa2e8b0723?

luizreis commented on includes/class-wc-payments-token-service.php:
```diff
@@ -133,46 +208,73 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
-			$token->delete();
+			// Do not remove tokens for the inactive customer.
+			if ( is_null( $inactive_customer_id ) || $token->get_meta( self::CUSTOMER_ID_META_KEY ) !== $inactive_customer_id ) {
+				$token->delete();
```
I'm not sure. I've just added the inactive mode customer check around the deletion.

@DanReyLop - is there a reason to not remove the deleted token from `$tokens`?

luizreis committed:
Catch import_customer_tokens exception

Catch and log import_customer_tokens exception to avoid
issues in the checkout and payment methods pages.

luizreis commented on includes/class-wc-payments-token-service.php:
```diff
@@ -103,6 +126,58 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
+	 *
+	 * @return array Token list with imported tokens.
+	 */
+	private function import_customer_tokens( $tokens, $customer_id, $user_id ) {
```
Great catch, I've added a catch block logging the error in https://github.com/Automattic/woocommerce-payments/pull/973/commits/34f9c165d49ae048f637d86f0f32b9d3eb51f17c to avoid breaking the checkout and payment methods pages.

dwainm committed:
Merge branch 'trunk' into add/965-associate-saved-card-with-customer

dwainm committed:
Merge branch 'trunk' into add/965-associate-saved-card-with-customer

dwainm commented:
@luizreis I have some testing feedback.

The following works fine:
✅ 1. Test accounts
✅ 2. Live accounts in test mode without the live customer

The next step doesn't work and I assume step 4 will not work either.

🔴 3. Live accounts in a test mode with live customer

Here's a video recording of what I find while testing:
https://drive.google.com/file/d/17xJL_KYkWaJxQYetTMQH1X1imSzWNZg5/view?usp=sharing

I've also shared the live account testing instructions internally.

luizreis commented:
This PR has been stale for a while and many things have changed since it was first introduced. Since we won't have the bandwidth necessary to work on it given our current priorities, I'll close this PR and unassign myself from the issue, so it can be reprioritized.
